File: media_kit_video/linux/opengl_anime4k_processor.cc
// This file is a part of media_kit
// (https://github.com/media-kit/media-kit).
//
// Copyright (c) 2026 Predidit.
// All rights reserved.
// Use of this source code is governed by MIT license that can be found in the
// LICENSE file.

#include "include/media_kit_video/opengl_anime4k_processor.h"

#include <array>
#include <cstring>
#include <iostream>
#include <vector>

#include "opengl_anime4k_restore_s_shaders.h"

#ifndef GL_HALF_FLOAT_OES
#define GL_HALF_FLOAT_OES 0x8D61
#endif

namespace {

constexpr GLfloat kVertices[] = {
    -1.0f, -1.0f, 0.0f, 0.0f,
     1.0f, -1.0f, 1.0f, 0.0f,
    -1.0f,  1.0f, 0.0f, 1.0f,
     1.0f,  1.0f, 1.0f, 1.0f,
};

bool HasExtension(const char* extensions, const char* extension) {
  return extensions != nullptr && std::strstr(extensions, extension) != nullptr;
}

bool SupportsES3() {
  const char* version =
      reinterpret_cast<const char*>(glGetString(GL_VERSION));
  return version != nullptr && std::strstr(version, "OpenGL ES 3") != nullptr;
}

GLuint CompileShader(GLenum type, const char* source) {
  GLuint shader = glCreateShader(type);
  glShaderSource(shader, 1, &source, nullptr);
  glCompileShader(shader);

  GLint compiled = GL_FALSE;
  glGetShaderiv(shader, GL_COMPILE_STATUS, &compiled);
  if (compiled == GL_TRUE) {
    return shader;
  }

  GLint log_length = 0;
  glGetShaderiv(shader, GL_INFO_LOG_LENGTH, &log_length);
  std::vector<char> log(log_length > 1 ? log_length : 1);
  glGetShaderInfoLog(shader, static_cast<GLsizei>(log.size()), nullptr,
                     log.data());
  std::cout << "media_kit: OpenGLAnime4KProcessor: shader compile failed: "
            << log.data() << std::endl;
  glDeleteShader(shader);
  return 0;
}

GLuint LinkProgram(const char* vertex_source, const char* fragment_source) {
  GLuint vertex_shader = CompileShader(GL_VERTEX_SHADER, vertex_source);
  if (vertex_shader == 0) {
    return 0;
  }

  GLuint fragment_shader = CompileShader(GL_FRAGMENT_SHADER, fragment_source);
  if (fragment_shader == 0) {
    glDeleteShader(vertex_shader);
    return 0;
  }

  GLuint program = glCreateProgram();
  glAttachShader(program, vertex_shader);
  glAttachShader(program, fragment_shader);
  glBindAttribLocation(program, 0, "a_position");
  glBindAttribLocation(program, 1, "a_uv");
  glLinkProgram(program);

  glDeleteShader(vertex_shader);
  glDeleteShader(fragment_shader);

  GLint linked = GL_FALSE;
  glGetProgramiv(program, GL_LINK_STATUS, &linked);
  if (linked == GL_TRUE) {
    return program;
  }

  GLint log_length = 0;
  glGetProgramiv(program, GL_INFO_LOG_LENGTH, &log_length);
  std::vector<char> log(log_length > 1 ? log_length : 1);
  glGetProgramInfoLog(program, static_cast<GLsizei>(log.size()), nullptr,
                      log.data());
  std::cout << "media_kit: OpenGLAnime4KProcessor: program link failed: "
            << log.data() << std::endl;
  glDeleteProgram(program);
  return 0;
}

}  // namespace

struct _OpenGLAnime4KProcessor {
  guint32 width = 0;
  guint32 height = 0;
  guint32 input_texture = 0;
  guint32 input_fbo = 0;
  std::array<guint32, 3> restore_textures = {0, 0, 0};
  std::array<guint32, 3> restore_fbos = {0, 0, 0};
  std::array<guint32, 4> programs = {0, 0, 0, 0};
  guint32 copy_program = 0;
  guint32 vertex_buffer = 0;
  bool shaders_ready = false;
  bool half_float_unavailable_reported = false;
};

OpenGLAnime4KProcessor* opengl_anime4k_processor_new() {
  return new OpenGLAnime4KProcessor();
}

void opengl_anime4k_processor_free(OpenGLAnime4KProcessor* self) {
  if (self == nullptr) {
    return;
  }

  if (self->input_texture != 0) {
    glDeleteTextures(1, &self->input_texture);
  }
  if (self->input_fbo != 0) {
    glDeleteFramebuffers(1, &self->input_fbo);
  }
  for (auto& texture : self->restore_textures) {
    if (texture != 0) {
      glDeleteTextures(1, &texture);
    }
  }
  for (auto& fbo : self->restore_fbos) {
    if (fbo != 0) {
      glDeleteFramebuffers(1, &fbo);
    }
  }
  for (auto& program : self->programs) {
    if (program != 0) {
      glDeleteProgram(program);
    }
  }
  if (self->copy_program != 0) {
    glDeleteProgram(self->copy_program);
  }
  if (self->vertex_buffer != 0) {
    glDeleteBuffers(1, &self->vertex_buffer);
  }

  delete self;
}

static gboolean ensure_shaders(OpenGLAnime4KProcessor* self) {
  if (self->shaders_ready) {
    return TRUE;
  }

  for (size_t i = 0; i < self->programs.size(); ++i) {
    self->programs[i] = LinkProgram(
        anime4k_restore_s_gl::kVertexShader,
        anime4k_restore_s_gl::kPassFragmentShaders[i]);
    if (self->programs[i] == 0) {
      return FALSE;
    }
  }

  self->copy_program = LinkProgram(anime4k_restore_s_gl::kVertexShader,
                                   anime4k_restore_s_gl::kCopyFragmentShader);
  if (self->copy_program == 0) {
    return FALSE;
  }

  glGenBuffers(1, &self->vertex_buffer);
  glBindBuffer(GL_ARRAY_BUFFER, self->vertex_buffer);
  glBufferData(GL_ARRAY_BUFFER, sizeof(kVertices), kVertices, GL_STATIC_DRAW);
  glBindBuffer(GL_ARRAY_BUFFER, 0);

  self->shaders_ready = true;
  return TRUE;
}

static gboolean create_texture_and_fbo(guint32 width,
                                       guint32 height,
                                       GLenum internal_format,
                                       GLenum format,
                                       GLenum type,
                                       guint32* texture,
                                       guint32* fbo) {
  glGenTextures(1, texture);
  glBindTexture(GL_TEXTURE_2D, *texture);
  glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_NEAREST);
  glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_NEAREST);
  glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, GL_CLAMP_TO_EDGE);
  glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, GL_CLAMP_TO_EDGE);
  glTexImage2D(GL_TEXTURE_2D, 0, internal_format, width, height, 0, format,
               type, nullptr);

  glGenFramebuffers(1, fbo);
  glBindFramebuffer(GL_FRAMEBUFFER, *fbo);
  glFramebufferTexture2D(GL_FRAMEBUFFER, GL_COLOR_ATTACHMENT0, GL_TEXTURE_2D,
                         *texture, 0);
  const GLenum status = glCheckFramebufferStatus(GL_FRAMEBUFFER);

  glBindFramebuffer(GL_FRAMEBUFFER, 0);
  glBindTexture(GL_TEXTURE_2D, 0);

  if (status == GL_FRAMEBUFFER_COMPLETE) {
    return TRUE;
  }

  glDeleteFramebuffers(1, fbo);
  glDeleteTextures(1, texture);
  *fbo = 0;
  *texture = 0;
  return FALSE;
}

static void delete_frame_textures(OpenGLAnime4KProcessor* self) {
  if (self->input_texture != 0) {
    glDeleteTextures(1, &self->input_texture);
    self->input_texture = 0;
  }
  if (self->input_fbo != 0) {
    glDeleteFramebuffers(1, &self->input_fbo);
    self->input_fbo = 0;
  }
  for (auto& texture : self->restore_textures) {
    if (texture != 0) {
      glDeleteTextures(1, &texture);
      texture = 0;
    }
  }
  for (auto& fbo : self->restore_fbos) {
    if (fbo != 0) {
      glDeleteFramebuffers(1, &fbo);
      fbo = 0;
    }
  }
  self->width = 0;
  self->height = 0;
}

gboolean opengl_anime4k_processor_ensure_size(OpenGLAnime4KProcessor* self,
                                               guint32 width,
                                               guint32 height) {
  if (self == nullptr) {
    return FALSE;
  }

  width = width > 0 ? width : 1;
  height = height > 0 ? height : 1;

  if (!ensure_shaders(self)) {
    return FALSE;
  }

  if (self->input_texture != 0 && self->input_fbo != 0 &&
      self->width == width && self->height == height) {
    return TRUE;
  }

  delete_frame_textures(self);

  if (!create_texture_and_fbo(width, height, GL_RGBA, GL_RGBA,
                              GL_UNSIGNED_BYTE, &self->input_texture,
                              &self->input_fbo)) {
    return FALSE;
  }

  GLenum half_internal_format = GL_RGBA;
  GLenum half_type = GL_HALF_FLOAT_OES;
  const char* extensions =
      reinterpret_cast<const char*>(glGetString(GL_EXTENSIONS));
  const bool supports_es3 = SupportsES3();
  if (supports_es3) {
#ifdef GL_RGBA16F
    half_internal_format = GL_RGBA16F;
#endif
#ifdef GL_HALF_FLOAT
    half_type = GL_HALF_FLOAT;
#endif
  } else if (!HasExtension(extensions, "GL_OES_texture_half_float") ||
             !HasExtension(extensions, "GL_EXT_color_buffer_half_float")) {
    if (!self->half_float_unavailable_reported) {
      std::cout << "media_kit: OpenGLAnime4KProcessor: half-float render "
                   "targets are unavailable."
                << std::endl;
      self->half_float_unavailable_reported = true;
    }
    delete_frame_textures(self);
    return FALSE;
  }

  for (size_t i = 0; i < self->restore_textures.size(); ++i) {
    if (!create_texture_and_fbo(width, height, half_internal_format, GL_RGBA,
                                half_type, &self->restore_textures[i],
                                &self->restore_fbos[i])) {
      std::cout << "media_kit: OpenGLAnime4KProcessor: restore framebuffer "
                << i << " is incomplete." << std::endl;
      delete_frame_textures(self);
      return FALSE;
    }
  }

  self->width = width;
  self->height = height;
  return TRUE;
}

guint32 opengl_anime4k_processor_get_input_fbo(
    OpenGLAnime4KProcessor* self) {
  return self != nullptr ? self->input_fbo : 0;
}

static gboolean render_pass(OpenGLAnime4KProcessor* self,
                            guint32 program,
                            guint32 output_fbo,
                            guint32 source_texture,
                            guint32 conv_texture) {
  if (program == 0 || output_fbo == 0 || source_texture == 0) {
    return FALSE;
  }

  GLint old_program = 0;
  GLint old_array_buffer = 0;
  GLint old_framebuffer = 0;
  GLint old_active_texture = 0;
  GLint old_viewport[4] = {0, 0, 0, 0};
  GLboolean blend_enabled = glIsEnabled(GL_BLEND);
  GLboolean depth_enabled = glIsEnabled(GL_DEPTH_TEST);
  GLboolean scissor_enabled = glIsEnabled(GL_SCISSOR_TEST);

  glGetIntegerv(GL_CURRENT_PROGRAM, &old_program);
  glGetIntegerv(GL_ARRAY_BUFFER_BINDING, &old_array_buffer);
  glGetIntegerv(GL_FRAMEBUFFER_BINDING, &old_framebuffer);
  glGetIntegerv(GL_ACTIVE_TEXTURE, &old_active_texture);
  glGetIntegerv(GL_VIEWPORT, old_viewport);

  glDisable(GL_BLEND);
  glDisable(GL_DEPTH_TEST);
  glDisable(GL_SCISSOR_TEST);
  glBindFramebuffer(GL_FRAMEBUFFER, output_fbo);
  glViewport(0, 0, self->width, self->height);
  glUseProgram(program);
  glBindBuffer(GL_ARRAY_BUFFER, self->vertex_buffer);

  glEnableVertexAttribArray(0);
  glEnableVertexAttribArray(1);
  glVertexAttribPointer(0, 2, GL_FLOAT, GL_FALSE, 4 * sizeof(GLfloat),
                        reinterpret_cast<const void*>(0));
  glVertexAttribPointer(1, 2, GL_FLOAT, GL_FALSE, 4 * sizeof(GLfloat),
                        reinterpret_cast<const void*>(2 * sizeof(GLfloat)));

  GLint size_location = glGetUniformLocation(program, "tex_size");
  if (size_location >= 0) {
    glUniform2f(size_location, static_cast<GLfloat>(self->width),
                static_cast<GLfloat>(self->height));
  }

  glActiveTexture(GL_TEXTURE0);
  glBindTexture(GL_TEXTURE_2D, source_texture);
  GLint source_location = glGetUniformLocation(program, "source_texture");
  if (source_location >= 0) {
    glUniform1i(source_location, 0);
  }

  if (conv_texture != 0) {
    glActiveTexture(GL_TEXTURE1);
    glBindTexture(GL_TEXTURE_2D, conv_texture);
    GLint conv_location = glGetUniformLocation(program, "conv_texture");
    if (conv_location >= 0) {
      glUniform1i(conv_location, 1);
    }
  }

  glDrawArrays(GL_TRIANGLE_STRIP, 0, 4);

  glDisableVertexAttribArray(0);
  glDisableVertexAttribArray(1);
  glActiveTexture(GL_TEXTURE1);
  glBindTexture(GL_TEXTURE_2D, 0);
  glActiveTexture(GL_TEXTURE0);
  glBindTexture(GL_TEXTURE_2D, 0);
  glUseProgram(old_program);
  glBindBuffer(GL_ARRAY_BUFFER, old_array_buffer);
  glBindFramebuffer(GL_FRAMEBUFFER, old_framebuffer);
  glViewport(old_viewport[0], old_viewport[1], old_viewport[2],
             old_viewport[3]);
  if (blend_enabled) {
    glEnable(GL_BLEND);
  }
  if (depth_enabled) {
    glEnable(GL_DEPTH_TEST);
  }
  if (scissor_enabled) {
    glEnable(GL_SCISSOR_TEST);
  }
  glActiveTexture(old_active_texture);

  return glGetError() == GL_NO_ERROR;
}

gboolean opengl_anime4k_processor_process(OpenGLAnime4KProcessor* self,
                                           guint32 output_fbo) {
  if (self == nullptr || self->input_texture == 0 || self->input_fbo == 0) {
    return FALSE;
  }

  if (!render_pass(self, self->programs[0], self->restore_fbos[0],
                   self->input_texture, 0)) {
    return FALSE;
  }
  if (!render_pass(self, self->programs[1], self->restore_fbos[1],
                   self->restore_textures[0], 0)) {
    return FALSE;
  }
  if (!render_pass(self, self->programs[2], self->restore_fbos[2],
                   self->restore_textures[1], 0)) {
    return FALSE;
  }
  return render_pass(self, self->programs[3], output_fbo, self->input_texture,
                     self->restore_textures[2]);
}

gboolean opengl_anime4k_processor_copy_input_to_fbo(
    OpenGLAnime4KProcessor* self,
    guint32 output_fbo) {
  if (self == nullptr || self->copy_program == 0 || self->input_texture == 0) {
    return FALSE;
  }
  return render_pass(self, self->copy_program, output_fbo, self->input_texture,
                     0);
}

File: media_kit_video/macos/media_kit_video/Sources/media_kit_video/plugin/OpenGLAnime4KProcessor.swift
// This file is a part of media_kit
// (https://github.com/media-kit/media-kit).
//
// Copyright (c) 2026 Predidit.
// All rights reserved.
// Use of this source code is governed by MIT license that can be found in the
// LICENSE file.

import CoreGraphics
import Foundation
import OpenGL.GL
import OpenGL.GL3

public final class OpenGLAnime4KProcessor {
  private let vertices: [GLfloat] = [
    -1.0, -1.0, 0.0, 0.0,
     1.0, -1.0, 1.0, 0.0,
    -1.0,  1.0, 0.0, 1.0,
     1.0,  1.0, 1.0, 1.0,
  ]

  private var width: GLsizei = 0
  private var height: GLsizei = 0
  private var inputTexture: GLuint = 0
  public private(set) var inputFrameBuffer: GLuint = 0
  private var restoreTextures: [GLuint] = [0, 0, 0]
  private var restoreFrameBuffers: [GLuint] = [0, 0, 0]
  private var programs: [GLuint] = [0, 0, 0, 0]
  private var copyProgram: GLuint = 0
  public var initFailed = false
  private var vertexBuffer: GLuint = 0

  deinit {
    deleteFrameTextures()
    deleteShaders()
  }

  public func ensureSize(_ size: CGSize) -> Bool {
    if initFailed {
      return false
    }

    let requiredWidth = max(GLsizei(size.width), 1)
    let requiredHeight = max(GLsizei(size.height), 1)

    if !ensureShaders() {
      initFailed = true
      return false
    }

    if inputTexture != 0 &&
      inputFrameBuffer != 0 &&
      width == requiredWidth &&
      height == requiredHeight
    {
      return true
    }

    deleteFrameTextures()

    guard let input = createTextureAndFrameBuffer(
      width: requiredWidth,
      height: requiredHeight,
      internalFormat: GLint(GL_RGBA),
      format: GLenum(GL_RGBA),
      type: GLenum(GL_UNSIGNED_BYTE)
    ) else {
      initFailed = true
      return false
    }

    inputTexture = input.texture
    inputFrameBuffer = input.frameBuffer

    for index in 0..<restoreTextures.count {
      var restore = createTextureAndFrameBuffer(
        width: requiredWidth,
        height: requiredHeight,
        internalFormat: GLint(GL_RGBA16F),
        format: GLenum(GL_RGBA),
        type: GLenum(GL_HALF_FLOAT)
      )

      if restore == nil {
        restore = createTextureAndFrameBuffer(
          width: requiredWidth,
          height: requiredHeight,
          internalFormat: GLint(GL_RGBA),
          format: GLenum(GL_RGBA),
          type: GLenum(GL_UNSIGNED_BYTE)
        )
      }

      guard let restoreResult = restore else {
        NSLog("OpenGLAnime4KProcessor: restore framebuffer \(index) is incomplete")
        deleteFrameTextures()
        initFailed = true
        return false
      }
      restoreTextures[index] = restoreResult.texture
      restoreFrameBuffers[index] = restoreResult.frameBuffer
    }

    width = requiredWidth
    height = requiredHeight
    return true
  }

  public func process(outputFrameBuffer: GLuint) -> Bool {
    if inputTexture == 0 || inputFrameBuffer == 0 {
      return false
    }

    if !renderPass(
      program: programs[0],
      outputFrameBuffer: restoreFrameBuffers[0],
      sourceTexture: inputTexture
    ) {
      return false
    }
    if !renderPass(
      program: programs[1],
      outputFrameBuffer: restoreFrameBuffers[1],
      sourceTexture: restoreTextures[0]
    ) {
      return false
    }
    if !renderPass(
      program: programs[2],
      outputFrameBuffer: restoreFrameBuffers[2],
      sourceTexture: restoreTextures[1]
    ) {
      return false
    }
    return renderPass(
      program: programs[3],
      outputFrameBuffer: outputFrameBuffer,
      sourceTexture: inputTexture,
      convTexture: restoreTextures[2]
    )
  }

  public func copyInput(to outputFrameBuffer: GLuint) -> Bool {
    if inputTexture == 0 || copyProgram == 0 {
      return false
    }

    return renderPass(
      program: copyProgram,
      outputFrameBuffer: outputFrameBuffer,
      sourceTexture: inputTexture
    )
  }

  private func ensureShaders() -> Bool {
    if initFailed {
      return false
    }
    if copyProgram != 0 && programs.allSatisfy({ $0 != 0 }) {
      return true
    }

    for index in 0..<Anime4KRestoreCNNShaders.restorePasses.count {
      let program = linkProgram(
        vertexSource: Anime4KRestoreCNNShaders.vertex,
        fragmentSource: Anime4KRestoreCNNShaders.restorePasses[index]
      )
      if program == 0 {
        initFailed = true
        return false
      }
      programs[index] = program
    }

    copyProgram = linkProgram(
      vertexSource: Anime4KRestoreCNNShaders.vertex,
      fragmentSource: Anime4KRestoreCNNShaders.copyFragment
    )
    if copyProgram == 0 {
      initFailed = true
      return false
    }

    glGenBuffers(1, &vertexBuffer)
    glBindBuffer(GLenum(GL_ARRAY_BUFFER), vertexBuffer)
    vertices.withUnsafeBytes { buffer in
      glBufferData(
        GLenum(GL_ARRAY_BUFFER),
        buffer.count,
        buffer.baseAddress,
        GLenum(GL_STATIC_DRAW)
      )
    }
    glBindBuffer(GLenum(GL_ARRAY_BUFFER), 0)
    return true
  }

  private func createTextureAndFrameBuffer(
    width: GLsizei,
    height: GLsizei,
    internalFormat: GLint,
    format: GLenum,
    type: GLenum
  ) -> (texture: GLuint, frameBuffer: GLuint)? {
    var texture: GLuint = 0
    glGenTextures(1, &texture)
    glBindTexture(GLenum(GL_TEXTURE_2D), texture)
    glTexParameteri(GLenum(GL_TEXTURE_2D), GLenum(GL_TEXTURE_MIN_FILTER), GL_NEAREST)
    glTexParameteri(GLenum(GL_TEXTURE_2D), GLenum(GL_TEXTURE_MAG_FILTER), GL_NEAREST)
    glTexParameteri(GLenum(GL_TEXTURE_2D), GLenum(GL_TEXTURE_WRAP_S), GL_CLAMP_TO_EDGE)
    glTexParameteri(GLenum(GL_TEXTURE_2D), GLenum(GL_TEXTURE_WRAP_T), GL_CLAMP_TO_EDGE)
    glTexImage2D(
      GLenum(GL_TEXTURE_2D),
      0,
      internalFormat,
      width,
      height,
      0,
      format,
      type,
      nil
    )

    var frameBuffer: GLuint = 0
    glGenFramebuffers(1, &frameBuffer)
    glBindFramebuffer(GLenum(GL_FRAMEBUFFER), frameBuffer)
    glFramebufferTexture2D(
      GLenum(GL_FRAMEBUFFER),
      GLenum(GL_COLOR_ATTACHMENT0),
      GLenum(GL_TEXTURE_2D),
      texture,
      0
    )

    let status = glCheckFramebufferStatus(GLenum(GL_FRAMEBUFFER))
    glBindFramebuffer(GLenum(GL_FRAMEBUFFER), 0)
    glBindTexture(GLenum(GL_TEXTURE_2D), 0)

    if status == GLenum(GL_FRAMEBUFFER_COMPLETE) {
      return (texture, frameBuffer)
    }

    glDeleteFramebuffers(1, &frameBuffer)
    glDeleteTextures(1, &texture)
    return nil
  }

  private func renderPass(
    program: GLuint,
    outputFrameBuffer: GLuint,
    sourceTexture: GLuint,
    convTexture: GLuint = 0
  ) -> Bool {
    if program == 0 || outputFrameBuffer == 0 || sourceTexture == 0 {
      return false
    }

    var initialError = glGetError()
    while initialError != GL_NO_ERROR {
      initialError = glGetError()
    }

    var oldProgram: GLint = 0
    var oldArrayBuffer: GLint = 0
    var oldVertexArray: GLint = 0
    var oldFrameBuffer: GLint = 0
    var oldActiveTexture: GLint = 0
    var oldViewport: [GLint] = [0, 0, 0, 0]
    let blendEnabled = glIsEnabled(GLenum(GL_BLEND))
    let depthEnabled = glIsEnabled(GLenum(GL_DEPTH_TEST))
    let scissorEnabled = glIsEnabled(GLenum(GL_SCISSOR_TEST))

    glGetIntegerv(GLenum(GL_CURRENT_PROGRAM), &oldProgram)
    glGetIntegerv(GLenum(GL_ARRAY_BUFFER_BINDING), &oldArrayBuffer)
    glGetIntegerv(GLenum(GL_VERTEX_ARRAY_BINDING), &oldVertexArray)
    glGetIntegerv(GLenum(GL_FRAMEBUFFER_BINDING), &oldFrameBuffer)
    glGetIntegerv(GLenum(GL_ACTIVE_TEXTURE), &oldActiveTexture)
    oldViewport.withUnsafeMutableBufferPointer { buffer in
      glGetIntegerv(GLenum(GL_VIEWPORT), buffer.baseAddress)
    }

    glDisable(GLenum(GL_BLEND))
    glDisable(GLenum(GL_DEPTH_TEST))
    glDisable(GLenum(GL_SCISSOR_TEST))
    glBindFramebuffer(GLenum(GL_FRAMEBUFFER), outputFrameBuffer)
    glViewport(0, 0, width, height)
    glUseProgram(program)

    var localVAO: GLuint = 0
    glGenVertexArrays(1, &localVAO)
    glBindVertexArray(localVAO)

    glBindBuffer(GLenum(GL_ARRAY_BUFFER), vertexBuffer)
    glEnableVertexAttribArray(0)
    glEnableVertexAttribArray(1)
    glVertexAttribPointer(
      0,
      2,
      GLenum(GL_FLOAT),
      GLboolean(GL_FALSE),
      GLsizei(4 * MemoryLayout<GLfloat>.size),
      UnsafeRawPointer(bitPattern: 0)
    )
    glVertexAttribPointer(
      1,
      2,
      GLenum(GL_FLOAT),
      GLboolean(GL_FALSE),
      GLsizei(4 * MemoryLayout<GLfloat>.size),
      UnsafeRawPointer(bitPattern: 2 * MemoryLayout<GLfloat>.size)
    )

    let texSizeLocation = glGetUniformLocation(program, "tex_size")
    if texSizeLocation >= 0 {
      glUniform2f(texSizeLocation, GLfloat(width), GLfloat(height))
    }

    glActiveTexture(GLenum(GL_TEXTURE0))
    glBindTexture(GLenum(GL_TEXTURE_2D), sourceTexture)
    let sourceLocation = glGetUniformLocation(program, "source_texture")
    if sourceLocation >= 0 {
      glUniform1i(sourceLocation, 0)
    }

    if convTexture != 0 {
      glActiveTexture(GLenum(GL_TEXTURE1))
      glBindTexture(GLenum(GL_TEXTURE_2D), convTexture)
      let convLocation = glGetUniformLocation(program, "conv_texture")
      if convLocation >= 0 {
        glUniform1i(convLocation, 1)
      }
    }

    glDrawArrays(GLenum(GL_TRIANGLE_STRIP), 0, 4)

    glDisableVertexAttribArray(0)
    glDisableVertexAttribArray(1)
    glBindBuffer(GLenum(GL_ARRAY_BUFFER), 0)
    glBindVertexArray(0)
    glDeleteVertexArrays(1, &localVAO)

    glActiveTexture(GLenum(GL_TEXTURE1))
    glBindTexture(GLenum(GL_TEXTURE_2D), 0)
    glActiveTexture(GLenum(GL_TEXTURE0))
    glBindTexture(GLenum(GL_TEXTURE_2D), 0)
    glUseProgram(GLuint(oldProgram))
    glBindBuffer(GLenum(GL_ARRAY_BUFFER), GLuint(oldArrayBuffer))
    glBindFramebuffer(GLenum(GL_FRAMEBUFFER), GLuint(oldFrameBuffer))
    glBindVertexArray(GLuint(oldVertexArray))
    glViewport(oldViewport[0], oldViewport[1], oldViewport[2], oldViewport[3])
    if blendEnabled == GL_TRUE {
      glEnable(GLenum(GL_BLEND))
    }
    if depthEnabled == GL_TRUE {
      glEnable(GLenum(GL_DEPTH_TEST))
    }
    if scissorEnabled == GL_TRUE {
      glEnable(GLenum(GL_SCISSOR_TEST))
    }
    glActiveTexture(GLenum(oldActiveTexture))

    let err = glGetError()
    if err != GL_NO_ERROR {
      NSLog("OpenGLAnime4KProcessor: renderPass failed with OpenGL error: 0x\(String(err, radix: 16))")
      return false
    }
    return true
  }

  private func linkProgram(vertexSource: String, fragmentSource: String) -> GLuint {
    let vertexShader = compileShader(type: GLenum(GL_VERTEX_SHADER), source: vertexSource)
    if vertexShader == 0 {
      return 0
    }

    let fragmentShader = compileShader(type: GLenum(GL_FRAGMENT_SHADER), source: fragmentSource)
    if fragmentShader == 0 {
      glDeleteShader(vertexShader)
      return 0
    }

    let program = glCreateProgram()
    glAttachShader(program, vertexShader)
    glAttachShader(program, fragmentShader)
    glBindAttribLocation(program, 0, "a_position")
    glBindAttribLocation(program, 1, "a_uv")
    glLinkProgram(program)
    glDeleteShader(vertexShader)
    glDeleteShader(fragmentShader)

    var linked: GLint = 0
    glGetProgramiv(program, GLenum(GL_LINK_STATUS), &linked)
    if linked == GL_TRUE {
      return program
    }

    logProgramError(program)
    glDeleteProgram(program)
    return 0
  }

  private func compileShader(type: GLenum, source: String) -> GLuint {
    let shader = glCreateShader(type)
    source.withCString { pointer in
      var sourcePointer: UnsafePointer<GLchar>? = pointer
      glShaderSource(shader, 1, &sourcePointer, nil)
      glCompileShader(shader)
    }

    var compiled: GLint = 0
    glGetShaderiv(shader, GLenum(GL_COMPILE_STATUS), &compiled)
    if compiled == GL_TRUE {
      return shader
    }

    logShaderError(shader)
    glDeleteShader(shader)
    return 0
  }

  private func logShaderError(_ shader: GLuint) {
    var logLength: GLint = 0
    glGetShaderiv(shader, GLenum(GL_INFO_LOG_LENGTH), &logLength)
    var log = [GLchar](repeating: 0, count: max(Int(logLength), 1))
    log.withUnsafeMutableBufferPointer { buffer in
      glGetShaderInfoLog(shader, GLsizei(buffer.count), nil, buffer.baseAddress)
    }
    let message = log.withUnsafeBufferPointer { buffer in
      String(cString: buffer.baseAddress!)
    }
    NSLog("OpenGLAnime4KProcessor: shader compile failed: \(message)")
  }

  private func logProgramError(_ program: GLuint) {
    var logLength: GLint = 0
    glGetProgramiv(program, GLenum(GL_INFO_LOG_LENGTH), &logLength)
    var log = [GLchar](repeating: 0, count: max(Int(logLength), 1))
    log.withUnsafeMutableBufferPointer { buffer in
      glGetProgramInfoLog(program, GLsizei(buffer.count), nil, buffer.baseAddress)
    }
    let message = log.withUnsafeBufferPointer { buffer in
      String(cString: buffer.baseAddress!)
    }
    NSLog("OpenGLAnime4KProcessor: program link failed: \(message)")
  }

  private func deleteFrameTextures() {
    if inputTexture != 0 {
      glDeleteTextures(1, &inputTexture)
      inputTexture = 0
    }
    if inputFrameBuffer != 0 {
      glDeleteFramebuffers(1, &inputFrameBuffer)
      inputFrameBuffer = 0
    }
    for index in 0..<restoreTextures.count {
      if restoreTextures[index] != 0 {
        var texture = restoreTextures[index]
        glDeleteTextures(1, &texture)
        restoreTextures[index] = 0
      }
      if restoreFrameBuffers[index] != 0 {
        var frameBuffer = restoreFrameBuffers[index]
        glDeleteFramebuffers(1, &frameBuffer)
        restoreFrameBuffers[index] = 0
      }
    }
    width = 0
    height = 0
  }

  private func deleteShaders() {
    for index in 0..<programs.count {
      if programs[index] != 0 {
        glDeleteProgram(programs[index])
        programs[index] = 0
      }
    }
    if copyProgram != 0 {
      glDeleteProgram(copyProgram)
      copyProgram = 0
    }
    if vertexBuffer != 0 {
      glDeleteBuffers(1, &vertexBuffer)
      vertexBuffer = 0
    }
  }
}
